src/aspire/source/coordinates.py

```diff
@@ -273,170 +300,119 @@ def _get_mrc_shapes(self):
-        optics = star["optics"]
-        micrographs = star["micrographs"]
 
-        # optics groups
-        optics_groups = [None]  # start indexing at 1
-        for _, row in optics.iterrows():
-            optics_groups.append(RlnOpticsGroup(row))
+        # merge DataFrames from CTF files
+        dfs = []
+        for f in ctf:
+            # ASPIRE's CTF Estimator produces legacy (=< 3.0) STAR files containing one row
+            star = Relion30StarFile(f)
+            dfs.append(star.get_block_by_index(0))
 
-        # micrographs
-        if not len(micrographs) == len(self.mrc_paths):
-            raise ValueError(
-                f"{ctf_starfile} has CTF information for {len(micrographs)}",
-                f" micrographs but this source has {len(self.mrc_paths)} micrographs.",
-            )
+        df = pd.concat(dfs, ignore_index=True)
 
-        for mrc_idx, row in micrographs.iterrows():
-            # extract parameters not in the optics groups
-            params = {}
-            params["defocus_u"], params["defocus_v"], params["defocus_angle"] = (
-                float(row._rlnDefocusU),
-                float(row._rlnDefocusV),
-                float(row._rlnDefocusAngle),
-            )
-            # get parameters from corresponding optics group
-            optics_group = optics_groups[int(row._rlnOpticsGroup)]
-            # set the rest of the CTF parameters
-            (
-                params["voltage"],
-                params["cs"],
-                params["amplitude_contrast"],
-                params["pixel_size"],
-            ) = (
-                optics_group.voltage,
-                optics_group.cs,
-                optics_group.amplitude_contrast,
-                optics_group.pixel_size,
-            )
-            self._process_each_ctf(params, mrc_idx)
+        self._extract_ctf(df)
 
-    def _populate_ctf_from_list(
-        self,
-        ctf_files,
-    ):
+    def import_relion_ctf(self, ctf):
         """
-        Populates CTF filters and metadata based on a list of .star files containing CTF parameters..
-        :param ctf_files: A list of .star files containing CTF parameters for micrographs.
-        (Note: number of files must match number of micrographs in CoordinateSource)
+        Import CTF information Relion micrograph STAR files containing CTF information.
+
+        :param ctf: Path to a Relion micrograph STAR file containing CTF information.
+            Note that number of files provided must match number of micrographs in this
+            `CoordinateSource`.
         """
-        if not len(ctf_files) == len(self.mrc_paths):
+        relion_version = getRelionStarFileVersion(ctf)
+        if not relion_version:
             raise ValueError(
-                "Number of CTF STAR files must match number of micrographs."
+                f"Cannot recognize {ctf} as a valid Relion STAR file containing micrograph information."
+            )
+        if relion_version == "3.0":
```

Comment:
This pattern happens twice in the PR, which makes me think that it might make sense to group them together in a simple wrapper class, `RelionStarFile` that does the version detection and has some method `get_data_block`, which does what we want.

Reply:
Done, let me know what you think. Thanks!

Reply:
Looks good!